### PR TITLE
Refactor events

### PR DESCRIPTION
**Pull request by c25vdw**
make it so that any classses that will need to talk with `socket` will have direct access to the socket. This removes a lot of callback binding inside `GameState` (there is none now), but in return, a lot of back and forth happens all over the places. 

It's still better imo, since now all events inside a class is type-constrained by the event definition:

Classes that are responsible for part of the system can define their own message type in `common/messages/newcomponent.ts`, and specify `Up` and `Down` events (server is at the cloud, so up is client2server). Corresponding classes in server and client can then define their socket type like `type MyMoreRestrictedSocket = Socket<UpEvents, DownEvents>;` and use that socket definition in their class, which prevents access to events belonging to other components, i.e. spectator cannot emit an event that is meant to be used by scoreboard, since its message type are different.

---
One caveat I can think of is that the socket on the client side (probably?) will accumulate listeners because the socket is now global. we could probably clean up all listeners when entering the start scene, but that is something to figure out later.

### CODE REVIEW

File: common/messages/gameStart.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+export interface DownEvents {
+  // updateScoreboard: (data: Array<{ color: string, hex: number, points: number }>) => void;
+  // endSequence: (data: Array<{ color: string, hex: number, points: number }>) => void;
+  // startSequence: () => void;
+  // showVotingSequence: (votingSequence: string) => void;
+  // hideVotingSequence: () => void;
+  // sendVotingCountdown: (secondsLeft: number) => void;
```
Comment (AdityaHarvi) [inline common/messages/gameStart.ts:7]: Are we able to delete the commented-out lines from this file? I noticed that you included them in the other new fiels.

File: common/messages/scoreboard.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import {TetrominoType} from "../TetrominoType"
+import {PlayerPosition, PlayerID, MoveEvent} from "../message"
+
+export type ColoredScore = { color: string, hex: number, points: number };
+
+export interface DownEvents {
+  updateScoreboard: (data: Array<ColoredScore>) => void;
+  endSequence: (data: Array<{ color: string, hex: number, points: number }>) => void;
```
Comment (AdityaHarvi) [inline common/messages/scoreboard.ts:8]: could use `Array<ColoredScore>` here as well

File: common/messages/spectator.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import {TetrominoType} from "../TetrominoType"
+import {PlayerPosition, PlayerID, MoveEvent} from "../message"
+
+export interface DownEvents {
+  endSequence: (data: Array<{ color: string, hex: number, points: number }>) => void;
```
Comment (AdityaHarvi) [inline common/messages/spectator.ts:5]: could use `Array<ColoredScore>` here as well

File: common/messages/sceneGame.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import {TetrominoType} from "../TetrominoType"
```
Comment (AdityaHarvi) [inline common/messages/sceneGame.ts:1]: Should this file be renamed to sceneGameArena to match up with the file naming scheme you have?

**Comment by c25vdw** [comment]
the `common/messages/gameStart.ts` is not yet used, probably because we are halfway building restarting of the game. the name of that file can change, or its content merged into one of the other scene message.

**Review by AdityaHarvi — APPROVED** [review]

**Pull request merged**